### ISSUE TITLE
Buggy display of Skirmish master entries

### ISSUE THREAD

**Issue by J-T-de** [Sh0wdown/UnofficialCrusaderPatch#650]
After a game in spectator mode, the greatest lord stats get displayed correctly, however if you check the stats later, the display is wrong.

For those who don't know, the statistics get saved in your documents folder in a file called skmaster2.dat (Crusader) or skmasterEx.dat (Crusader Extreme). I found out that the results are stored correctly (so you can extract the correct results), but get displayed wrongly afterwards, but the match-up is stored wrong.

The solution would be to find out why the matchup is wrong. There will definitely remain some artifacts in the preview (player1 is hardcoded to be a human, battles won't sort properly as the score for player1 is not computed correctly), but the results are correct. Reference for the file format: https://github.com/sourcehold/sourcehold-maps/blob/master/structure/kaitai/skmaster_dat.ksy
![result](https://user-images.githubusercontent.com/50496721/93238440-ba7fc900-f781-11ea-9593-a18f92fc3499.PNG)
![ingame](https://user-images.githubusercontent.com/50496721/93238454-bd7ab980-f781-11ea-9dcc-22c718b7475d.PNG)

This is kinda related to https://github.com/Sh0wdown/UnofficialCrusaderPatch/issues/648, because this issue also touches the rendering of the greatest lord screen.

**Comment by patel-nikhil** [comment]
> however if you check the stats later, the display is wrong

The screenshot helps, but can you summarize what is displayed wrong so it’s a bit easier to see? At a glance I see for example two AIs in the aiId list and two in the greatest lord screen - is it that the player numbers and sorting are wrong or is there more than that?

> but get displayed wrongly afterwards, but the match-up is stored wrong
So after it is shown wrongly in game, the file gets overwritten with the new (wrong) stats?

**Comment by J-T-de** [comment]
I was not yet able to find out what technically is the problem, or even how exactly this problem occurs.

In the example:
- the actual matchup was what is written in the `playerNames` array, i.e., 4 pigs and 1 wolf.
- what is rendered in the greates lord screen is when the `aiId` is not 0 (there is also an `isActive` array, but I did not yet test which one is responsible for this faulty behaviour

The data is inconsistent, there should be `aiId[2]=PIG, aiId[3]=PIG, aiId[4]=PIG`. Iirc, this "bug" occurs always if the red player is exchanged for an ai (who would have thought) and the following ai has the same id.

(The file format linked is kinda outdated, the current version is here: https://github.com/sourcehold/sourcehold-data-library/blob/master/stronghold_crusader/structs/skmaster_dat.ksy) Each entry has the form
```
 entry:
    seq:
      - id: description
        type: description
      - id: preview_data
        type: preview_data
      - id: results
        type: s1125         # imported struct
```
Where the description encodes which map/mission is played, the preview data basically describes the matchup and the results are what is displayed in the greatest lord screen. I am pretty sure, that the `preview_data` gets initialized in the wrong way.

**Comment by GRhin** [comment]
It sounds like the problem is the skmasters file, not the display. Wherever the skmasters file gets written, those other ais are not written as active players.
Is this a spectator mode problem, maybe even active spectator map problems? Or does this happen with regular games as well?

**Comment by J-T-de** [comment]
Yes, I think it is a problem with the skirmishmaster file, because at the end of a spectator game, the stats are correctly rendered. (But due to the faulty file, it gets displayed wrong when you open the results of passed games).

With regular games I could not observe this bug, I am 99.9% sure that this bug is introduced with the spectator mode. I did not use the active spectator mode, so I am quite sure it is a problem of the "normal" spectator mode (even though I believe the active spectator mode will have the same issue).
